text/0021-any-and-all-operators.md

```diff
@@ -0,0 +1,165 @@
+
+## Motivation
+
+The `all` and `any` operators support a common use case: iterating over a set and checking if a predicate holds for any or all members of the set.
+
+Cedar currently lacks a general way to express this use case. The only alternative is to expand the `all` or `any` checks into conjunctions or disjunctions. For example:
+
+```
+// Using all
+["a", "ab", "abc"].all(it like "a*")
+
+// Using &&
+("a" like "a*") &&
+("ab" like "a*") &&
+("abc" like "a*")
+```
+
+This manual expansion works for literal sets but cannot be applied to non-literal sets like `context.portNumbers`.
+
+The new operators provide a built-in way to concisely express these universal and existential quantifications over a set.
+
+## Detailed design
+
+The Cedar syntax, semantics, and type system are all extended to support the new operators. These extensions are backward compatible, and existing policies won't be affected.
+
+### Extending the syntax
+
+This RFC proposes extending the Cedar grammar as follows:
+
+```
+Access    :=
+  '.' QUANTIFIER '(' Predicate ')'
+  | '.' IDENT ['(' [ExprList] ')']
+  | '[' STR ']'
+Quantifier := 'all' | 'any'
+Predicate  := ...
+```
+
+The `Predicate` expression differs from `Expr` in two ways:
+
+1. Predicates can use the keyword `it`. A `Predicate` defines the body of function with a single parameter, and we use `it` as the parameter name. This is similar to [Kotlin's `it` keyword](https://kotlinlang.org/docs/lambdas.html#it-implicit-name-of-a-single-parameter).
+
+2. Predicates cannot contain nested `any` or `all` expressions. We disallow nested quantifiers for both performance and analyzability reasons. Nested quantifiers allow writing expressions that run in $O(2^n)$ time, where $n$ is the number of nested quantifiers. For example:
+
+```
+[0,1].all(
+  [0,1].all(
+    ...
+      [0,1].all(true)...))
+```
+
+Because nested quantifiers are prohibited, there is no need to support predicates with arbitrary parameter names. The `it` keyword sufficiently extends the grammar.
+
+### Extending the semantics
+
+The semantics of `all` and `any` are straightforward when there are no errors:
+- `E.all(P)` evaluates to true when `P[e/it]` is true for every element `e` in `E`.
+- `E.any(P)` evaluates to true when `P[e/it]` is true for some element `e` in `E`.
+
+But what happens if evaluating `P` errors on some element `e`? In this case, the evaluation of `all` or `any` terminates abruptly with a distinguished `QuantifierError`.
+
+The `QuantifierError` error may include static information (such as source location), but no dynamic information (such as the value that caused the error to be thrown). This error handling approach ensures that evaluating  `all` and `any` is _deterministic_: the result is the same regardless of iteration order.
+
+For example, consider the expression `[1, true].all(it like "foo")`. Regardless of the order in which the predicate is applied, the expression returns the same `QuantifierError`.
```

Comment:
Add:

> Compare this to the alternative semantics that just propagates the error discovered, when it is discovered. Then evaluating this expression could lead to the error that "a Long was found where a String was expected" or it could lead to the error "A Boolean was found where a String was expected." Which one depends on evaluation order, but no clear order exists because set elements are specifically unordered.